src/names/applications.cpp

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2021 Jeremy Rand
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <names/applications.h>
+
+#include <names/encoding.h>
+
+#include <regex>
+
+namespace
+{
+
+NameNamespace
+NamespaceFromString (const std::string& str)
+{
+    if (str == "d/")
+        return NameNamespace::Domain;
+    if (str == "dd/")
+        return NameNamespace::DomainData;
+    if (str == "id/")
+        return NameNamespace::Identity;
+    if (str == "idd/")
+        return NameNamespace::IdentityData;
+
+    return NameNamespace::NonStandard;
+}
+
+std::string
+NamespaceToString (NameNamespace ns)
+{
+    switch (ns)
+    {
+        case NameNamespace::Domain:
+            return "d/";
+        case NameNamespace::DomainData:
+            return "dd/";
+        case NameNamespace::Identity:
+            return "id/";
+        case NameNamespace::IdentityData:
+            return "idd/";
+        case NameNamespace::NonStandard:
+            return "";
+    }
+
+    /* NameNamespace values are only ever created internally in the binary
+    and not received externally (except as string).  Thus it should never
+    happen (and if it does, is a severe bug) that we see an unexpected
+    value.  */
+    assert (false);
+}
+
+NameNamespace
+PurportedNamespaceFromName (const std::string& name)
+{
+    const auto slashPos = name.find("/");
+    if (slashPos == std::string::npos)
+    {
+        return NameNamespace::NonStandard;
+    }
+
+    const std::string namespaceStr = name.substr(0, slashPos+1);
+    return NamespaceFromString(namespaceStr);
+}
+
+} // anonymous namespace
+
+NameNamespace
+NamespaceFromName (const std::string& name)
+{
+    NameNamespace purported = PurportedNamespaceFromName(name);
+    const std::string purportedStr = NamespaceToString(purported);
+    const auto purportedLen = purportedStr.length();
+
+    switch (purported)
+    {
+        case NameNamespace::Domain:
+        {
+            const std::string label = name.substr(purportedLen);
+
+            if (!IsStringValid (label, NameEncoding::ASCII))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            if (label.length() < 1)
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            // Source: https://github.com/namecoin/proposals/blob/master/ifa-0001.md#keys
+            if (label.length() > 63)
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            // Source: https://github.com/namecoin/proposals/blob/master/ifa-0001.md#keys
+            std::regex domainPattern("^(xn--)?[a-z0-9]+(-[a-z0-9]+)*$");
+            if (!std::regex_match (label, domainPattern))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            // Reject digits-only labels
```

Comment:
Out of curiosity, why is this invalid?

Reply:
Same spec reference as the above two checks; I'll add a comment clarifying this.